### DISCUSSION TITLE
[Support] controller控制器接收的body参数，是否不能自动深度转换为自定义的复杂类的实例?还是需要一些配置才可以？

### DISCUSSION THREAD

**Discussion by DannyWin** [midwayjs/midway#4219] [Q&A] [answered]
### Describe the problem（描述问题）

代码：
`
export class UserDTO {
  id: number;
  uid: string;
  pwd: string;
  name: string;
  phone: string;

  getUid() {
    return this.uid;
  }
}

export class UserAddReq {
  list: UserDTO[];
  getList() {
    return this.list;
  }
}

@Post('/addUsers')
  async addUsers(@Body() param: UserAddReq) {
    const res = await this.userService.addUsers(param.list);
   console.log('param',param instanceof UserAddReq);
   console.log('param.list',param.list[0]  instanceof UserDTO)
    console.log('getList', userParam.getList());
    console.log('getUid', userParam.list[0].getUid());
    return res;
  }
`
这里分别打印param，param.list，getList，getUid
param显示true
param.list显示false
getList显示传入的list数据
getUid会报内部错误，userParam.list[0].getUid is not a function"
也就是说param被自动转换为UserAddReq类的实例，但是param.list不是UserDTO的实例，直接打印param.list，显示就是一个普通数组，没有类型信息，所以打印userParam.list[0].getUid()会报错

根据官方文档[controller#请求参数类型转换](https://www.midwayjs.org/docs/controller#%E8%AF%B7%E6%B1%82%E5%8F%82%E6%95%B0%E7%B1%BB%E5%9E%8B%E8%BD%AC%E6%8D%A2)这一节，写了“如果是复杂类型，如果指定的类型是 Class，将会自动转换为该类的实例。”，并且给出了示例，但是示例比较简单,如下：
`
// class
class UserDTO {
  name: string;

  getName() {
    return this.name;
  }
}

@Get('/')
async getUser(@Query() query: UserDTO): Promise<User> {
  // query.getName()
}
`

问题：controller控制器接收的body参数，是否不能自动深度转换为自定义的复杂类的实例，只能自动转换最外面的一层？还是需要有一些设置才能深度转换？

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(3.17.3)
✓ @midwayjs/bootstrap(3.17.1)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/busboy(not installed)
✓ @midwayjs/cache-manager(3.17.1)
✓ @midwayjs/captcha(3.17.1)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.17.1)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(3.17.3)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.17.3)
✓ @midwayjs/jwt(3.17.3)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.17.1)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(3.17.3)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(3.17.1)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(3.17.1)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.17.3)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.17.1)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by czy88840616** [comment]
文档下面有写，写interface就不会自动转了。要么不转，要么全转

**Reply by DannyWin** [reply]
谢谢大佬回复，我的意思是我想全转，但是转换不完全，body参数只有最外面的对象转换了，对象的属性（值是一个数组<每个元素是另一个类型>）是数组的情况时，数组中的元素没有自动转换
简单来说就是使用Class后，body参数确实被转换为Class A的实例，但是Class A的一个字段是数组，数组的每个元素类型是Class B，数组的元素没有被自动转换为Class B
请问是我这边的配置问题，还是框架设定就是这样

**Comment by DannyWin** [comment]
@czy88840616 谢谢大佬回复，我的意思是我想全转，但是转换不完全，body参数只有最外面的对象转换了，对象的属性（值是一个数组<每个元素是另一个类型>）是数组的情况时，数组中的元素没有自动转换
简单来说就是使用Class后，body参数确实被转换为Class A的实例，但是Class A的一个字段是数组，数组的每个元素类型是Class B，数组的元素没有被自动转换为Class B
请问是我这边的配置问题，还是框架设定就是这样

**Reply by czy88840616** [reply]
举个例子？我猜测跟 joi 配置有关，如果某个字段规则没写对或者没写，就不会转了

**Comment by DannyWin** [comment]
@czy88840616 代码如下，去掉了参数验证：
类型：
`
export class IconDTO {
  id: number;
  name: string;
  size: number;
  url: string;
  color: string;
  desc: string;
}

export class IconReq {
  list: IconDTO[];
}
`
控制器:
`
 @Post('/addIcons')
  async addIcons(@Body() param: IconReq) {
    console.log('IconReq', param instanceof IconReq);
    console.log('IconDTO', param.list[0] instanceof IconDTO);
    console.log('param', param);
  }
`
![image](https://github.com/user-attachments/assets/28fc4635-fe1e-4cf7-85f7-64ba50e8a1da)

**Reply by czy88840616** [reply]
这里的转换行为有多层，如果写了 Rule，则转换交给 validate 组件（joi），如果没有用，那么会交给 class-transformer。

**Reply by DannyWin** [reply]
好的，谢谢大佬回复，我这边不管写不写Rule，上例中的list都不能自动被转换为IconDTO的实例数组，我这边手动安装了class-transformer，添加@Type装饰器，然后可以自动深度转换了
IconReq类增加@Type装饰器：
`
export class IconReq {

  @Type(() => IconDTO)
  
  list: IconDTO[];
  
}
`
![image](https://github.com/user-attachments/assets/442ce9fa-e174-44d8-9770-7c5062227c51)

但是还是不清楚为什么madway没有自动调用class-transformer进行深度转换，如果确实是@Type装饰器必须要手动添加，感觉可以在文档中提一下

**Reply by czy88840616** [reply, marked as answer]
自动转换类这个逻辑太隐晦了，新版本 v4 中已经移除了，后续自动转换只会显式在 validate 时进行。

**Reply by DannyWin** [reply]
明白了，感谢大佬指点
